File: reference/java-producer-low-latency-azure/src/main/java/org/fsi/kafka/producer/FsiDlqHandler.java
package org.fsi.kafka.producer;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.header.internals.RecordHeader;
import org.apache.kafka.common.header.internals.RecordHeaders;
import org.apache.kafka.common.serialization.ByteArraySerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * FSI C4E DLQ Handler -- Java.
 *
 * Routes failed messages to {source-topic}.dlq with error categorization,
 * exponential backoff retry, and Kafka header metadata.
 *
 * Error classification:
 *   - Non-retryable (serialization, auth): sent directly to DLQ
 *   - Retryable (broker timeout, leader election): retried up to 3 times
 *     with exponential backoff (1s, 2s, 4s) before DLQ routing
 *
 * This is a reference implementation. Copy and adapt for your application.
 */
public class FsiDlqHandler implements AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(FsiDlqHandler.class);

    /** Maximum retry attempts before routing to DLQ. */
    public static final int MAX_RETRIES = 3;

    /** Base backoff in milliseconds (doubles each retry: 1s, 2s, 4s). */
    public static final long BASE_BACKOFF_MS = 1000;

    /** Exception types that should NOT be retried. */
    private static final Set<Class<? extends Exception>> NON_RETRYABLE = new HashSet<>(Arrays.asList(
            org.apache.kafka.common.errors.SerializationException.class,
            org.apache.kafka.common.errors.InvalidRecordException.class,
            org.apache.kafka.common.errors.RecordTooLargeException.class,
            org.apache.kafka.common.errors.TopicAuthorizationException.class,
            org.apache.kafka.common.errors.GroupAuthorizationException.class
    ));

    private final KafkaProducer<byte[], byte[]> dlqProducer;
    private final String dlqTopic;
    private final String sourceTopic;
    private final String clientId;
    private final AtomicLong dlqSent = new AtomicLong(0);

    /**
     * Create a DLQ handler with a dedicated raw-bytes producer.
     *
     * @param config Map with bootstrap.servers and optional SASL config.
     * @param sourceTopic Original topic name -- DLQ topic will be {sourceTopic}.dlq.
     */
    public FsiDlqHandler(Map<String, String> config, String sourceTopic) {
        this.sourceTopic = sourceTopic;
        this.dlqTopic = sourceTopic + ".dlq";
        this.clientId = config.getOrDefault("client.id", "fsi-dlq") + "-dlq";

        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
                config.getOrDefault("bootstrap.servers", "kafka.fsi.internal:9092"));
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
        props.put(ProducerConfig.CLIENT_ID_CONFIG, clientId);

        // Copy SASL config if present
        if (config.containsKey("security.protocol")) {
            props.put("security.protocol", config.get("security.protocol"));
        } else if (config.containsKey("sasl.jaas.config")) {
            props.put("security.protocol", "SASL_SSL");
        }
        if (config.containsKey("sasl.mechanism")) {
            props.put("sasl.mechanism", config.get("sasl.mechanism"));
        } else if (config.containsKey("sasl.jaas.config")) {
            props.put("sasl.mechanism", "PLAIN");
        }
        if (config.containsKey("sasl.jaas.config")) {
            props.put("sasl.jaas.config", config.get("sasl.jaas.config"));
        }

        this.dlqProducer = new KafkaProducer<>(props);
        log.info("DLQ handler initialized. DLQ topic: {}", dlqTopic);
    }

    /**
     * Check if an exception is retryable.
     *
     * @param e The exception to classify.
     * @return true if retryable, false if it should go directly to DLQ.
     */
    public boolean isRetryable(Exception e) {
        return !NON_RETRYABLE.stream().anyMatch(cls -> cls.isInstance(e));
    }

    /**
     * Calculate exponential backoff delay for a given retry count.
     *
     * @param retryCount Current retry attempt (0-based).
     * @return Backoff duration in milliseconds (1s, 2s, 4s).
     */
    public long getBackoffMs(int retryCount) {
        return BASE_BACKOFF_MS * (1L << retryCount);
    }

    /**
     * Send a failed message to the DLQ topic with error metadata headers.
     *
     * @param key Original message key as bytes (may be null).
     * @param value Original message value as bytes (may be null).
     * @param error The exception that caused the failure.
     * @param retryCount Number of retry attempts before this DLQ send.
     */
    public void sendToDlq(byte[] key, byte[] value, Exception error, int retryCount) {
        String errorType = classifyError(error);
        String errorMessage = error.getMessage() != null
                ? error.getMessage().substring(0, Math.min(error.getMessage().length(), 1024))
                : "null";
        String timestamp = Instant.now().toString();

        RecordHeaders headers = new RecordHeaders();
        headers.add(new RecordHeader("dlq.original.topic", sourceTopic.getBytes(StandardCharsets.UTF_8)));
        headers.add(new RecordHeader("dlq.error.type", errorType.getBytes(StandardCharsets.UTF_8)));
        headers.add(new RecordHeader("dlq.error.message", errorMessage.getBytes(StandardCharsets.UTF_8)));
        headers.add(new RecordHeader("dlq.timestamp", timestamp.getBytes(StandardCharsets.UTF_8)));
        headers.add(new RecordHeader("dlq.retry.count", String.valueOf(retryCount).getBytes(StandardCharsets.UTF_8)));
        headers.add(new RecordHeader("dlq.producer.client.id", clientId.getBytes(StandardCharsets.UTF_8)));

        ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(dlqTopic, null, key, value, headers);

        dlqProducer.send(record, (metadata, exception) -> {
            if (exception != null) {
                log.error("Failed to deliver to DLQ {}: {}", dlqTopic, exception.getMessage());
            } else {
                log.warn("Sent to DLQ {}-{} offset={} [error_type={}, retries={}]",
                        metadata.topic(), metadata.partition(), metadata.offset(),
                        errorType, retryCount);
            }
        });

        dlqSent.incrementAndGet();
    }

    /**
     * Classify an exception into a human-readable error category.
     *
     * @param e The exception to classify.
     * @return One of: SERIALIZATION, AUTH_DENIED, BROKER_TIMEOUT, UNKNOWN.
     */
    public String classifyError(Exception e) {
        if (e instanceof org.apache.kafka.common.errors.SerializationException
                || e instanceof org.apache.kafka.common.errors.InvalidRecordException
                || e instanceof org.apache.kafka.common.errors.RecordTooLargeException) {
            return "SERIALIZATION";
        }
        if (e instanceof org.apache.kafka.common.errors.TopicAuthorizationException
                || e instanceof org.apache.kafka.common.errors.GroupAuthorizationException
                || e instanceof org.apache.kafka.common.errors.ClusterAuthorizationException) {
            return "AUTH_DENIED";
        }
        if (e instanceof org.apache.kafka.common.errors.TimeoutException
                || e instanceof org.apache.kafka.common.errors.NotLeaderOrFollowerException
                || e instanceof org.apache.kafka.common.errors.LeaderNotAvailableException) {
            return "BROKER_TIMEOUT";
        }
        return "UNKNOWN";
    }

    /**
     * Get the total number of messages sent to DLQ.
     *
     * @return DLQ message count.
     */
    public long getDlqSent() {
        return dlqSent.get();
    }

    @Override
    public void close() {
        log.info("Closing DLQ handler. Total DLQ messages: {}", dlqSent.get());
        dlqProducer.flush();
        dlqProducer.close();
    }
}

File: reference/python-producer-low-latency-azure/fsi_dlq_handler.py
"""FSI C4E DLQ Handler -- Python.

Routes failed messages to {source-topic}.dlq with error categorization,
exponential backoff retry, and Kafka header metadata.

This is a reference implementation. Copy and adapt for your application.
"""

import logging
import random
import time
from datetime import datetime, timezone

from confluent_kafka import KafkaError, Producer

log = logging.getLogger(__name__)

# Error codes that should NOT be retried -- send directly to DLQ
NON_RETRYABLE_ERRORS = {
    KafkaError._VALUE_SERIALIZATION,
    KafkaError._KEY_SERIALIZATION,
    KafkaError._VALUE_DESERIALIZATION,
    KafkaError._KEY_DESERIALIZATION,
    KafkaError.TOPIC_AUTHORIZATION_FAILED,
    KafkaError.GROUP_AUTHORIZATION_FAILED,
    KafkaError.CLUSTER_AUTHORIZATION_FAILED,
}

# Maximum retry attempts before routing to DLQ
MAX_RETRIES = 3

# Base backoff in seconds (doubles each retry: 1s, 2s, 4s)
BASE_BACKOFF_S = 1


class FsiDlqHandler:
    """Routes failed messages to a Dead Letter Queue topic.

    Error categorization:
      - Non-retryable (serialization, auth): sent directly to DLQ
      - Retryable (broker timeout, leader election): retried up to 3 times
        with exponential backoff before DLQ routing

    DLQ topic naming follows {source-topic}.dlq convention.
    """

    def __init__(self, producer_config: dict, source_topic: str):
        """Create a DLQ handler with a dedicated producer.

        Args:
            producer_config: Dict with bootstrap.servers and optional SASL config.
            source_topic: Original topic name -- DLQ topic will be {source_topic}.dlq.
        """
        self._source_topic = source_topic
        self._dlq_topic = f"{source_topic}.dlq"
        self._client_id = producer_config.get(
            "client.id", "fsi-dlq"
        ) + "-dlq"

        # Build DLQ producer config -- minimal, just needs connectivity
        dlq_config = {
            "bootstrap.servers": producer_config.get(
                "bootstrap.servers", "localhost:9092"
            ),
            "client.id": self._client_id,
        }

        # Copy SASL config if present
        if producer_config.get("sasl.username"):
            dlq_config["security.protocol"] = "SASL_SSL"
            dlq_config["sasl.mechanism"] = "PLAIN"
            dlq_config["sasl.username"] = producer_config["sasl.username"]
            dlq_config["sasl.password"] = producer_config.get(
                "sasl.password", ""
            )

        self._producer = Producer(dlq_config)

        # Counter for observable metrics
        self._dlq_sent = 0

        log.info("DLQ handler initialized. DLQ topic: %s", self._dlq_topic)

    def handle_error(
        self,
        key: bytes,
        value_bytes: bytes,
        error,
        retry_count: int = 0,
    ):
        """Handle a failed message by classifying and routing.

        Args:
            key: Original message key as bytes (or None).
            value_bytes: Original message value as bytes (or None).
            error: The KafkaError or Exception that caused the failure.
            retry_count: Current retry attempt (0-based).

        Returns:
            None if message was sent to DLQ.
            Tuple of ("retry", backoff_seconds) if caller should retry.
        """
        error_code = getattr(error, "code", lambda: None)()

        # Non-retryable errors go directly to DLQ
        if error_code in NON_RETRYABLE_ERRORS:
            log.warning(
                "Non-retryable error on %s: %s -- routing to DLQ",
                self._source_topic,
                error,
            )
            self._send_to_dlq(key, value_bytes, error, retry_count)
            return None

        # Retryable errors: retry with backoff up to MAX_RETRIES
        if retry_count < MAX_RETRIES:
            backoff = BASE_BACKOFF_S * (2 ** retry_count) + random.uniform(
                0, 0.1 * (2 ** retry_count)
            )
            log.info(
                "Retryable error on %s (attempt %d/%d), backoff %.2fs: %s",
                self._source_topic,
                retry_count + 1,
                MAX_RETRIES,
                backoff,
                error,
            )
            return ("retry", backoff)

        # Exhausted retries -- route to DLQ
        log.warning(
            "Exhausted %d retries on %s: %s -- routing to DLQ",
            MAX_RETRIES,
            self._source_topic,
            error,
        )
        self._send_to_dlq(key, value_bytes, error, retry_count)
        return None

    def _send_to_dlq(self, key, value_bytes, error, retry_count):
        """Produce message to DLQ topic with error metadata headers."""
        error_type = self._classify_error(error)
        now = datetime.now(timezone.utc).isoformat()

        headers = {
            "dlq.original.topic": self._source_topic.encode("utf-8"),
            "dlq.error.type": error_type.encode("utf-8"),
            "dlq.error.message": str(error)[:1024].encode("utf-8"),
            "dlq.timestamp": now.encode("utf-8"),
            "dlq.retry.count": str(retry_count).encode("utf-8"),
            "dlq.producer.client.id": self._client_id.encode("utf-8"),
        }

        self._producer.produce(
            topic=self._dlq_topic,
            key=key,
            value=value_bytes,
            headers=headers,
            on_delivery=self._dlq_delivery_callback,
        )
        self._producer.poll(0)
        self._dlq_sent += 1

        log.warning(
            "Sent to DLQ %s [error_type=%s, retries=%d]",
            self._dlq_topic,
            error_type,
            retry_count,
        )

    def _dlq_delivery_callback(self, err, msg):
        """Callback for DLQ produce -- log failures but don't recurse."""
        if err is not None:
            log.error(
                "Failed to deliver to DLQ %s: %s",
                self._dlq_topic,
                err,
            )

    def _classify_error(self, error) -> str:
        """Map error to a human-readable category.

        Categories:
          SERIALIZATION -- key/value serialization or deserialization failure
          SCHEMA_INCOMPATIBLE -- schema registry compatibility rejection
          BROKER_TIMEOUT -- broker unreachable or request timed out
          AUTH_DENIED -- topic, group, or cluster authorization failure
          UNKNOWN -- unrecognized error
        """
        error_code = getattr(error, "code", lambda: None)()

        if error_code in {
            KafkaError._VALUE_SERIALIZATION,
            KafkaError._KEY_SERIALIZATION,
            KafkaError._VALUE_DESERIALIZATION,
            KafkaError._KEY_DESERIALIZATION,
        }:
            return "SERIALIZATION"

        if error_code in {
            KafkaError.TOPIC_AUTHORIZATION_FAILED,
            KafkaError.GROUP_AUTHORIZATION_FAILED,
            KafkaError.CLUSTER_AUTHORIZATION_FAILED,
        }:
            return "AUTH_DENIED"

        if error_code in {
            KafkaError._TIMED_OUT,
            KafkaError._MSG_TIMED_OUT,
            KafkaError.REQUEST_TIMED_OUT,
            KafkaError.NOT_LEADER_FOR_PARTITION,
            KafkaError.LEADER_NOT_AVAILABLE,
        }:
            return "BROKER_TIMEOUT"

        # Check for schema incompatibility by message text
        error_str = str(error).lower()
        if "schema" in error_str and (
            "incompatible" in error_str or "compatibility" in error_str
        ):
            return "SCHEMA_INCOMPATIBLE"

        return "UNKNOWN"

    @property
    def dlq_sent(self) -> int:
        """Total messages routed to DLQ."""
        return self._dlq_sent

    def close(self):
        """Flush and close the DLQ producer."""
        log.info(
            "Closing DLQ handler. Total DLQ messages: %d", self._dlq_sent
        )
        self._producer.flush(timeout=10)
